semantic/constraints.go

```diff
@@ -96,7 +96,11 @@ func (v ConstraintGenerator) typeof(n Node) (PolyType, error) {
 		ftv := n.ExternType.freeVars(nil)
 		subst := make(Substitution, len(ftv))
 		for _, tv := range ftv {
-			subst[tv] = v.cs.f.Fresh()
+			f := v.cs.f.Fresh()
+			for ftv.contains(f) {
+				f = v.cs.f.Fresh()
+			}
+			subst[tv] = f
```

Comment:
It should be possible to create a test case that enter the loop. That would be a good enough test case for me. Have a look at the extern type inference test cases that already exist. Since you can just pick the type variables the extern type use, you should be able to create a conflict that requires this loop to fix.